### DISCUSSION TITLE
Poll: Should new warnings emitted given the same previous contents be considered a breaking change?

### DISCUSSION THREAD

**Discussion by pawamoy** [mkdocs/mkdocs#3822] [Polls]
Should new warnings (log messages with WARNING level) emitted by MkDocs plugins given the same previous contents (user code/docs didn't change) be considered a breaking change?

Such warnings make strict builds fail, but strict builds are optional: it is still possible to build and deploy docs. This is similar to developers who opt into making their pytest runs fail when deprecation warnings are emitted and captured: even though the pytest run fails, we don't consider the deprecation warning itself to be a breaking change, and only the test runs are impacted, not the production software. Still, maybe it's kinder to your users to create transition periods before the warnings take effect? 

(Previously posted as a discussion here: https://github.com/mkdocs/mkdocs/discussions/3818)

**Comment by pawamoy** [comment]
I changed my vote to "Depends".

Depends because IMO we simply enter "public API" territory, and what can be considered stable or subject to unnotified changes in a project. Each project ultimately decides what is its public API, and what is not. Depends on the warnings as well: are they easy to get rid of, i.e. is it easy to "fix" the code or docs that the warnings point to? Are a lot of projects impacted?

**Reply by cdwilson** [reply]
(Depends)

FYI, Brett Cannon wrote a blog post about this exact scenario:

https://snarky.ca/why-i-dont-like-semver/

I don't think there is a canonical answer to this poll (one might even be able to make the argument that newly added warnings are "features", although I think that's a probably a stretch).

Also from https://hynek.me/articles/semver-will-not-save-you/:

> Because that’s all SemVer is: **a TL;DR of the changelog**.
> 
> What it does mean though, is that you can’t rely on the semantic meaning of SemVer and **you must treat every update as potentially breaking**.

Said another way, SemVer is just a maintainer's best-effort attempt at summarizing the nature of the changes, but the [spec is ambiguous](https://github.com/semver/semver/issues/468) in a lot of cases. If it was me, I'd just bump the minor version, document the change in the CHANGELOG, and move on.

**Reply by Andre601** [reply]
> FYI, Brett Cannon wrote a blog post about this exact scenario:
> 
> https://snarky.ca/why-i-dont-like-semver/

While off-topic, I would like to add the argument, that semver may force you to keep some old API you really don't want, unless you force a new major API version for nothing other than removing these old features.
I myself have this issue, where an API I have now has countless deprecated methods. I want to remove them, but at the same time don't want to make a new major release just for that, as Semver would require me to do so, given it's a breaking change... And there is really no other major changes that would be somewhat breaking for me, that I could do here.
It's quite a frustrating situation at times and not fun to deal with, yet so many people consider this to really be a good aproach for a version scheme.

Also, one rather interesting aproach on versions I saw was with the org.json library for Java, where new versions are nothing more than `yyyymmdd`, which is a rather strange aproach at versioning, but also helps in seeing how long the last update has been.

**Reply by pawamoy** [reply]
@cdwilson yeah I'm really starting to feel Brett's article. From one project to another, the definition of "API" and what part of it is public and/or stable can vary, so ultimately the numbers in a semver version are just a best effort to match what your users may expect.

**Reply by pawamoy** [reply]
@Andre601 yep, CalVer is used too in the Python ecosystem, for example by pip 😊
